// @grafbase/sdk/dist/src/connector/header.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Headers = exports.Header = void 0;
/**
 * Header used in connector calls.
 */
var Header = /** @class */ (function () {
    function Header(name, value) {
        this.name = name;
        this.value = value;
    }
    Header.prototype.toString = function () {
        return "{ name: \"".concat(this.name, "\", value: \"").concat(this.value, "\" }");
    };
    return Header;
}());
exports.Header = Header;
/**
 * An accumulator class to gather headers for a connector which supports
 * introspection headers.
 */
var Headers = /** @class */ (function () {
    function Headers() {
        this._headers = [];
        this._introspectionHeaders = [];
    }
    Object.defineProperty(Headers.prototype, "headers", {
        /**
         * All headers used in client requests.
         */
        get: function () {
            return this._headers;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Headers.prototype, "introspectionHeaders", {
        /**
         * All headers used in introspection requests.
         */
        get: function () {
            return this._introspectionHeaders;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Creates a header used in client requests.
     *
     * @param name - The name of the header
     * @param value - The value of the header
     */
    Headers.prototype.static = function (name, value) {
        this.headers.push(new Header(name, value));
    };
    /**
     * Creates a header used in introspection requests.
     *
     * @param name - The name of the header
     * @param value - The value of the header
     */
    Headers.prototype.introspection = function (name, value) {
        this.introspectionHeaders.push(new Header(name, value));
    };
    return Headers;
}());
exports.Headers = Headers;
