// @grafbase/sdk/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.auth = exports.connector = exports.config = exports.g = void 0;
var grafbase_schema_1 = require("./grafbase-schema");
var config_1 = require("./config");
var openapi_1 = require("./connector/openapi");
var graphql_1 = require("./connector/graphql");
var openid_1 = require("./auth/openid");
var jwt_1 = require("./auth/jwt");
var jwks_1 = require("./auth/jwks");
var dotenv_1 = __importDefault(require("dotenv"));
dotenv_1.default.config();
/**
 * A builder for a Grafbase schema definition.
 */
exports.g = new grafbase_schema_1.GrafbaseSchema();
/**
 * A constructor for a complete Grafbase configuration.
 */
function config(input) {
    return new config_1.Config(input);
}
exports.config = config;
exports.connector = {
    /**
     * Create a new OpenAPI connector object.
     *
     * @param params - The configuration parameters.
     */
    OpenAPI: function (params) {
        return new openapi_1.PartialOpenAPI(params);
    },
    /**
     * Create a new GraphQL connector object.
     *
     * @param params - The configuration parameters.
     */
    GraphQL: function (params) {
        return new graphql_1.PartialGraphQLAPI(params);
    }
};
exports.auth = {
    /**
     * Create a new OpenID authenticator.
     *
     * @param params - The configuration parameters.
     */
    OpenIDConnect: function (params) {
        return new openid_1.OpenIDAuth(params);
    },
    /**
     * Create a new JWT authenticator.
     *
     * @param params - The configuration parameters.
     */
    JWT: function (params) {
        return new jwt_1.JWTAuth(params);
    },
    /**
     * Create a new JWKS authenticator.
     *
     * @param params - The configuration parameters.
     */
    JWKS: function (params) {
        return new jwks_1.JWKSAuth(params);
    }
};
