// @grafbase/sdk/dist/src/typedefs/reference.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceDefinition = void 0;
var list_1 = require("./list");
var auth_1 = require("./auth");
var resolver_1 = require("./resolver");
var ReferenceDefinition = /** @class */ (function () {
    function ReferenceDefinition(referencedType) {
        this.referencedType = referencedType.name;
        this.isOptional = false;
    }
    /**
     * Set the field optional.
     */
    ReferenceDefinition.prototype.optional = function () {
        this.isOptional = true;
        return this;
    };
    /**
     * Allow multiple scalars to be used as values for the field.
     */
    ReferenceDefinition.prototype.list = function () {
        return new list_1.ListDefinition(this);
    };
    /**
     * Set the field-level auth directive.
     *
     * @param rules - A closure to build the authentication rules.
     */
    ReferenceDefinition.prototype.auth = function (rules) {
        return new auth_1.AuthDefinition(this, rules);
    };
    /**
     * Attach a resolver function to the field.
     *
     * @param name - The name of the resolver function file without the extension or directory.
     */
    ReferenceDefinition.prototype.resolver = function (name) {
        return new resolver_1.ResolverDefinition(this, name);
    };
    ReferenceDefinition.prototype.toString = function () {
        var required = this.isOptional ? '' : '!';
        return "".concat(this.referencedType).concat(required);
    };
    return ReferenceDefinition;
}());
exports.ReferenceDefinition = ReferenceDefinition;
