// @grafbase/sdk/dist/src/typedefs/resolver.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolverDefinition = void 0;
var auth_1 = require("./auth");
var cache_1 = require("./cache");
var ResolverDefinition = /** @class */ (function () {
    function ResolverDefinition(field, resolver) {
        this.field = field;
        this.resolver = resolver;
        this.requiresFields = null;
        this._arguments = {};
    }
    /**
     * Set the field-level auth directive.
     *
     * @param rules - A closure to build the authentication rules.
     */
    ResolverDefinition.prototype.auth = function (rules) {
        return new auth_1.AuthDefinition(this, rules);
    };
    /**
     * Set the field-level cache directive.
     *
     * @param params - The cache definition parameters.
     */
    ResolverDefinition.prototype.cache = function (params) {
        return new cache_1.CacheDefinition(this, new cache_1.FieldLevelCache(params));
    };
    /**
     * Declares that this resolver requires certain fields to function correctly
     *
     * @param fields - The fields this resolver requires
     */
    ResolverDefinition.prototype.requires = function (fields) {
        this.requiresFields = fields;
        return this;
    };
    /**
     * Adds arguments to this field that will be available in the resolver
     *
     * @param args - The arguments for this field
     */
    ResolverDefinition.prototype.arguments = function (args) {
        this._arguments = args;
        return this;
    };
    Object.defineProperty(ResolverDefinition.prototype, "allArguments", {
        get: function () {
            return __assign(__assign({}, this._arguments), this.field.allArguments);
        },
        enumerable: false,
        configurable: true
    });
    ResolverDefinition.prototype.toString = function () {
        var requires = this.requiresFields == null
            ? ''
            : " @requires(fields: \"".concat(this.requiresFields, "\")");
        return "".concat(this.field, " @resolver(name: \"").concat(this.resolver, "\")").concat(requires);
    };
    return ResolverDefinition;
}());
exports.ResolverDefinition = ResolverDefinition;
