// @grafbase/sdk/dist/src/typedefs/resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResolverDefinition = void 0;
var auth_1 = require("./auth");
var cache_1 = require("./cache");
var ResolverDefinition = /** @class */ (function () {
    function ResolverDefinition(field, resolver) {
        this.field = field;
        this.resolver = resolver;
    }
    /**
     * Set the field-level auth directive.
     *
     * @param rules - A closure to build the authentication rules.
     */
    ResolverDefinition.prototype.auth = function (rules) {
        return new auth_1.AuthDefinition(this, rules);
    };
    /**
     * Set the field-level cache directive.
     *
     * @param params - The cache definition parameters.
     */
    ResolverDefinition.prototype.cache = function (params) {
        return new cache_1.CacheDefinition(this, new cache_1.FieldLevelCache(params));
    };
    ResolverDefinition.prototype.toString = function () {
        return "".concat(this.field, " @resolver(name: \"").concat(this.resolver, "\")");
    };
    return ResolverDefinition;
}());
exports.ResolverDefinition = ResolverDefinition;
